envoy/config/filter/http/jwt_authn/v2/config.proto

```diff
@@ -0,0 +1,239 @@
+  //
+  // If no explicit location is specified, the following default locations are
+  // tried in order:
+  //
+  // 1. The Authorization header using the Bearer schema.
+  //    See `here <https://tools.ietf.org/html/rfc6750#section-2.1>`_
+  //    Example::
+  //
+  //       Authorization: Bearer <token>.
+  //
+  // 2. `access_token` query parameter.
+  //    See `this <https://tools.ietf.org/html/rfc6750#section-2.3>`_
+  //
+
+  // Multiple JWTs can be verified for a request. Each JWT has to be extracted
+  // from the locations its issuer specified or from the default locations.
+
+  // Specify the HTTP headers to extract JWT token.
+  //
+  // For examples, following config::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_headers:
+  //   - name: x-goog-iap-jwt-assertion
+  //
+  // can be used to extract token from header::
+  //
+  //   x-goog-iap-jwt-assertion: <JWT>.
+  //
+  repeated JwtHeader jwt_headers = 7;
+
+  // JWT is sent in a query parameter. `jwt_params` represents the
+  // query parameter names.
+  //
+  // For example, if config is::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_params:
+  //   - jwt_token
+  //
+  // The JWT format in query parameter is::
+  //
+  //    /path?jwt_token=<JWT>
+  //
+  repeated string jwt_params = 8;
+
+  // This field specifies the header name to forward a successfully verified
+  // JWT payload to the backend. The forwarded data is::
+  //
+  //    base64_encoded(jwt_payload_in_JSON)
+  //
+  // If it is not specified, the payload will not be forwarded.
+  // If multiple JWT payloads needed to be forwarded, distinct header names
+  // are required.
+  string jwt_payload_header = 9;
+}
+
+// This message specifies how to fetch JWKS from remote and how to cache it.
+message RemoteJwks {
+  // The HTTP URI to fetch the JWKS.
+  //
+  // Example:
+  //
+  // .. code-block:: yaml
+  //
+  //    http_uri:
+  //    - uri: https://www.googleapis.com/oauth2/v1/certs
+  //      cluster: jwt.www.googleapis.com|443
+  //
+  envoy.api.v2.core.HttpUri http_uri = 1;
+
+  // Duration after which the cached JWKS should be expired. If not
+  // specified, default cache duration is 5 minutes.
+  google.protobuf.Duration cache_duration = 2;
+}
+
+// This message specifies a header location to extract JWT token.
+message JwtHeader {
+  // The HTTP header name.
+  string name = 1 [(validate.rules).string.min_bytes = 1];
+
+  // The value prefix. The value format is "value_prefix<token>"
+  // For example, for "Authorization: Bearer <token>",
+  // value_prefix="Bearer " with a space at the end.
+  string value_prefix = 2;
+}
+
+// This is the Envoy HTTP filter config for JWT authentication.
+// [#not-implemented-hide:]
+message JwtAuthentication {
+  // List of JWT rules to valide.
+  repeated JwtRule jwt_rules = 1;
+
+  // If true, the request is allowed if JWT is missing.
+  // Default is false, a request without JWT is not allowed.
+  bool allow_missing_jwt = 2;
```

Comment:
Bring back old comment: I'm not quite understand your intention for this flag and how it works together with the allow_failed_jwt. I think the other one (allow_failed_jwt) is not needed.

Reply:
allow_failed_jwt is at JwtRule level.  It allows users to config each type of JWT to allow it to fail or not.
Are you saying that we just don't need this flag at all, neither at JwtRule level nor at global level?

Reply:
Let's remove `allow_failed_jwt` for now. Likely we will need something at JwtRule level, but likely we need more option than a bool (e.g. redirect when unauthenticated), we can defer this to later.

Reply:
OK, will remove it